storage/storage.rs

```diff
@@ -276,165 +144,124 @@ impl<D> MVCCStorage<D> {
+            .map_err(|error| Durability { name: self.name.to_owned(), source: error })?;
+
+        if conflict.is_none() {
+            Self::apply_commit(commit_sequence_number, &self.keyspaces, &self.isolation_manager)
+                .map_err(|error| Keyspace { name: self.name.to_owned(), source: Arc::new(error) })?;
+        }
+
+        Self::persist_commit_status(conflict.is_none(), commit_sequence_number, &self.durability_service)
+            .map_err(|error| Durability { name: self.name.clone(), source: error })?;
+
+        match conflict {
+            Some(conflict) => Err(StorageCommitError::Isolation { name: self.name.clone(), conflict }),
+            None => Ok(()),
+        }
+    }
+
+    fn set_initial_put_status(&self, snapshot: &impl CommittableSnapshot<Durability>)
+    where
+        Durability: DurabilityService,
+    {
+        for buffer in snapshot.operations() {
             let writes = buffer.writes().write().unwrap();
             let puts = writes.iter().filter_map(|(key, write)| match write {
-                Write::Put { value, reinsert } => Some((key, value, reinsert)),
+                Write::Put { value, reinsert, known_to_exist } => Some((key, value, reinsert, *known_to_exist)),
                 _ => None,
             });
-            for (key, value, reinsert) in puts {
+            for (key, value, reinsert, known_to_exist) in puts {
                 let wrapped = StorageKeyReference::new_raw(buffer.keyspace_id, ByteReference::new(key.bytes()));
-                let existing_stored: Option<Option<ByteArray<BUFFER_VALUE_INLINE>>> = self
-                    .get(wrapped, commit_record.open_sequence_number(), |reference| {
-                        // Only copy if the value is the same
-                        (reference.bytes() == value.bytes()).then(|| ByteArray::from(reference))
-                    })
-                    .unwrap(); // TODO
-                reinsert.store(existing_stored.flatten().is_none(), Ordering::Release);
+                if known_to_exist {
+                    reinsert.store(false, Ordering::Release);
+                } else {
+                    let existing_stored = self
+                        .get::<BUFFER_VALUE_INLINE>(wrapped, snapshot.open_sequence_number())
+                        .unwrap() // TODO
+                        .is_some_and(|reference| reference.bytes() == value.bytes());
+                    reinsert.store(!existing_stored, Ordering::Release);
+                }
             }
         }
+    }
 
-        //  1. make durable and get sequence number
-        let commit_sequence_number =
-            self.durability_service.sequenced_write(&commit_record).map_err(|err| MVCCStorageError {
-                storage_name: self.name.to_string(),
-                kind: MVCCStorageErrorKind::DurabilityError { source: err },
-            })?;
-
-        // 2,3,4:
-        self.try_write_commit_record(
+    fn apply_commit(
```

Comment:
Can we try to make this take the data?

Reply:
First pass: let's construct the write batches before sending the record to isolation manager.

Reply:
Let's then inline this function